domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,96 @@
+package org.oppia.android.domain.question
+
+import org.oppia.android.app.model.FractionGrade
+import org.oppia.android.app.model.UserAssessmentPerformance
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Private mutable class that stores a grade as a fraction.
+ */
+private class MutableFractionGrade(var numerator: Double, var denominator: Double)
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset.
+ */
+@Singleton
+internal
+class QuestionAssessmentCalculation @Inject constructor(
+  @ViewHintPenalty private val viewHintPenalty: Double,
+  @WrongAnswerPenalty private val wrongAnswerPenalty: Double,
+  @MaxScorePerQuestion private val maxScorePerQuestion: Double
+) {
+  private lateinit var skillIdList: List<String>
+  private lateinit var questionSessionMetrics: List<QuestionSessionMetrics>
+  private lateinit var totalScore: MutableFractionGrade
+  private lateinit var scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+
+  /** Initialize member fields. */
+  internal fun initialize(
+    skillIdList: List<String>,
+    questionSessionMetrics: List<QuestionSessionMetrics>
+  ) {
+    this.skillIdList = skillIdList
+    this.questionSessionMetrics = questionSessionMetrics
+    this.totalScore = MutableFractionGrade(0.0, questionSessionMetrics.size.toDouble())
+    createScorePerSkillMapping()
+  }
+
+  /** Initialize the scorePerSkillMapping member field. */
+  private fun createScorePerSkillMapping() {
+    scorePerSkillMapping = mutableMapOf()
+    for (skillId in skillIdList) {
+      scorePerSkillMapping[skillId] = MutableFractionGrade(0.0, 0.0)
+    }
+  }
+
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
+    for (questionMetric in questionSessionMetrics) {
+      val totalHintsPenalty = questionMetric.numberOfHintsUsed * viewHintPenalty
+      val totalWrongAnswerPenalty =
+        (questionMetric.numberOfAnswersSubmitted - 1) * wrongAnswerPenalty
+      var questionScore = 0.0
+      if (!questionMetric.didViewSolution) {
+        questionScore =
+          maxOf(maxScorePerQuestion - totalHintsPenalty - totalWrongAnswerPenalty, questionScore)
+      }
+      this.totalScore.numerator += questionScore
+      for (linkedSkillId in questionMetric.question.linkedSkillIdsList) {
+        if (!scorePerSkillMapping.containsKey(linkedSkillId)) continue
+        scorePerSkillMapping[linkedSkillId]!!.numerator += questionScore
```

Comment:
There could be cases where the score per skill mapping has a numerator and a denominator as 0. Are we handling that case anywhere? Ideally it’s best to remove those entries lower in the stack so that when we actually divide the numerator and denominator we do not need to handle that scenario

Reply:
Addressed! Thanks for catching that

Reply:
Maybe I missed this..won't we still have cases where both the numerator and denominator can be 0?

Reply:
We're removing any skills with a 0 numerator/denominator in the computeAll() method